### PR TITLE
Fixed toolchain spec packing for hosts with RPM < 4.18.

### PR DESCRIPTION
**Pull request by PawelWMS**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [X] The toolchain/worker package manifests are up-to-date
- [X] Any updated packages successfully build (or no packages were changed)
- [X] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [X] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [X] All package sources are available
- [X] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [X] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [X] All source files have up-to-date hashes in the `*.signatures.json` files
- [X] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [X] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary 

Toolchain packages are being packed into SRPMs using the host's RPM (Azure Linux worker chroot is not available at that time). For RPM versions < 4.18 (common on Ubuntu 22.04 and earlier) the only valid `%patch` macro usage is either `%patch[number]` or `%patch -P [number]`. Since the former is deprecated starting from RPM 4.18, I've updated toolchain specs to use `-P [number]`.

###### Change Log  

- Updated toolchain specs' `%patch` macros.
- Enabled missing 2 patches in `perl.spec`.
- Extended the spec PR check to prevent invalid `%patch` macro usage for toolchain packages.

###### Does this affect the toolchain?  

Yes.

###### Associated issues  

- #7136

###### Test Methodology

- Local tests of the PR check and package builds.
- [Full toolchain build](https://dev.azure.com/mariner-org/mariner/_build/results?buildId=506999&view=results).

### CODE REVIEW

File: SPECS/zip/zip.spec [hunk 1]
```diff
@@ -31,6 +30,9 @@ make prefix=%{buildroot}/%{_prefix} MANDIR=%{buildroot}/usr/share/man/man1 -f un
 %{_mandir}/*
 
 %changelog
+*   Thu Feb 15 2024 Pawel Winogrodzki <pawelwi@microsoft.com> - 3.0-6
+-   - Updated patch application macros.
```
Comment (cwize1) [inline SPECS/zip/zip.spec:34]: nit: Extra `-`

File: SPECS/perl/perl.spec [hunk 2]
```diff
@@ -6847,6 +6838,10 @@ popd
 
 # Old changelog entries are preserved in CVS.
 %changelog
+* Thu Feb 15 2024 Pawel Winogrodzki <pawelwi@microsoft.com> - 4:5.38.2-504
+- Updated patch application macros.
+- Applying missed patches 14 and 202.
```
Comment (cwize1) [inline SPECS/perl/perl.spec:6843]: Do we know why these were excluded?

File: toolkit/scripts/check_spec_guidelines.py [hunk 3]
```diff
@@ -121,7 +126,31 @@ def check_source_attribution(spec_path: str):
     return False
 
 
-def check_spec(spec_path):
+def check_toolchain_patch_lines(spec_path: str, toolchain_specs: list):
```
Comment (cwize1) [inline toolkit/scripts/check_spec_guidelines.py:129]: Should `toolchain_specs` be `set` not `list`?

File: SPECS/perl/perl.spec (hunk 2 above)
Comment (PawelWMS) [reply SPECS/perl/perl.spec:6843]: Yes, I spoke with @nicogbg. Turns out we accidentally missed them during an upgrade but there was no reason to skip them.

**Review by cwize1 — APPROVED** [review]

**Pull request merged**